File: src/main/groovy/net/zomis/brainf/analyze/BrainfuckAnalyzer.java
package net.zomis.brainf.analyze;

import groovy.transform.CompileStatic;
import net.zomis.brainf.model.BrainfuckRunner;
import net.zomis.brainf.model.ast.tree.Syntax;

@CompileStatic
public interface BrainfuckAnalyzer {

    default Object createMemoryData() { return null; }
    default void beforeStart(BrainfuckRunner runner) { }
    default void after(Brainalyze analyze, BrainfuckRunner runner) { }

    default void print() { }

    default void beforePerform(MemoryCell cell, BrainfuckRunner runner, Syntax command) { }
    default void afterPerform(MemoryCell cell, BrainfuckRunner runner, Syntax command) { }

    default void beforeWhile(MemoryCell cell, BrainfuckRunner runner) {}
    default void afterWhile(MemoryCell cell, BrainfuckRunner runner) {}
    default void beforeEndWhile(MemoryCell cell, BrainfuckRunner runner) {}
    default void afterEndWhile(MemoryCell cell, BrainfuckRunner runner) {}

}

File: brainduck-kotlin/src/commonMain/kotlin/net/zomis/brainf/analyze/BrainfuckAnalyzer.kt
package net.zomis.brainf.analyze;

import net.zomis.brainf.model.BrainfuckRunner
import net.zomis.brainf.model.ast.tree.Syntax

interface BrainfuckAnalyzer {

    fun createMemoryData(): Any {
        return Unit
    }
    fun beforeStart(runner: BrainfuckRunner) { }
    fun after(analyze: Brainalyze, runner: BrainfuckRunner) { }
    fun print() { }
    fun beforePerform(cell: MemoryCell, runner: BrainfuckRunner, command: Syntax) { }
    fun afterPerform(cell: MemoryCell, runner: BrainfuckRunner, command: Syntax) { }

    fun beforeWhile(cell: MemoryCell, runner: BrainfuckRunner) {}
    fun afterWhile(cell: MemoryCell, runner: BrainfuckRunner) {}
    fun beforeEndWhile(cell: MemoryCell, runner: BrainfuckRunner) {}
    fun afterEndWhile(cell: MemoryCell, runner: BrainfuckRunner) {}

}
